drivers/fpga/fpga_ice40.c

```diff
@@ -0,0 +1,600 @@
+ * CS polarity is normal (active low). Zephyr's SPI driver model currently
+ * does not handle these types of quirks (in contrast to e.g. Linux).
+ *
+ * The logical alternative would be to put the CS into GPIO mode, perform 3
+ * separate SPI transfers (inverting CS polarity as necessary) and then
+ * restore the default pinctrl settings. On some higher-end microcontrollers
+ * and microprocessors, it's possible to do that without breaking the iCE40
+ * timing requirements.
+ *
+ * However, on lower-end microcontrollers, the amount of time that elapses
+ * between SPI transfers does break the iCE40 timing requirements. That
+ * leaves us with the bitbanging option. Of course, on lower-end
+ * microcontrollers, the amount of time required to execute something
+ * like gpio_pin_configure_dt() dwarfs the 2*500 nanoseconds needed to
+ * achieve the minimum 1 MHz clock rate for loading the iCE40 bistream. So
+ * in order to bitbang on lower-end microcontrollers, we actually require
+ * direct register access to the set and clear registers.
+ *
+ * With that, this driver is left with 2 possible modes of operation which
+ * are:
+ * - FPGA_ICE40_LOAD_WITH_SPI (for higher-end microcontrollers)
+ * - FPGA_ICE40_LOAD_WITH_GPIO (for lower-end microcontrollers)
+ */
+
+#ifndef BITS_PER_NIBBLE
+#define BITS_PER_NIBBLE 4
+#endif
+
+#ifndef BITS_PER_BYTE
+#define BITS_PER_BYTE 8
+#endif
+
+#ifndef NIBBLES_PER_BYTE
+#define NIBBLES_PER_BYTE (BITS_PER_BYTE / BITS_PER_NIBBLE)
+#endif
+
+#define FPGA_ICE40_SPI_FREQ_MIN 1000000
+#define FPGA_ICE40_SPI_FREQ_MAX 25000000
+
+#define FPGA_ICE40_CRESET_DELAY_NS_MIN 200
+#define FPGA_ICE40_CONFIG_DELAY_US_MIN 300
+#define FPGA_ICE40_LEADING_CLOCKS_MIN  8
+#define FPGA_ICE40_TRAILING_CLOCKS_MIN 49
+
+#define FPGA_ICE40_LOAD_MODE_SPI  0
+#define FPGA_ICE40_LOAD_MODE_GPIO 1
+
+LOG_MODULE_REGISTER(fpga_ice40);
+
+struct fpga_ice40_data {
+	uint32_t crc;
+	/* simply use crc32 as info */
+	char info[2 * sizeof(uint32_t) + 1];
+	bool on;
+	bool loaded;
+	struct k_spinlock lock;
+};
+
+struct fpga_ice40_config {
+	struct spi_dt_spec bus;
+	struct gpio_dt_spec cdone;
+	struct gpio_dt_spec creset;
+	struct gpio_dt_spec clk;
+	struct gpio_dt_spec pico;
+	volatile gpio_port_pins_t *set;
+	volatile gpio_port_pins_t *clear;
+	uint16_t mhz_delay_count;
+	uint8_t creset_delay_ns;
+	uint16_t config_delay_us;
+	uint8_t leading_clocks;
+	uint8_t trailing_clocks;
+	fpga_api_load load;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif
+};
+
+static void fpga_ice40_crc_to_str(uint32_t crc, char *s)
+{
+	char ch;
+	uint8_t i;
+	uint8_t nibble;
+	const char *table = "0123456789abcdef";
+
+	for (i = 0; i < sizeof(crc) * NIBBLES_PER_BYTE; ++i, crc >>= BITS_PER_NIBBLE) {
+		nibble = crc & GENMASK(BITS_PER_NIBBLE, 0);
+		ch = table[nibble];
+		s[sizeof(crc) * NIBBLES_PER_BYTE - i - 1] = ch;
+	}
+
+	s[sizeof(crc) * NIBBLES_PER_BYTE] = '\0';
+}
+
+static inline void fpga_ice40_delay(size_t n)
+{
+	for (; n > 0; --n) {
+	}
+}
```

Comment:
Could you explain why this is needed? It seems you are setting `n` to `0` anyway? If you just need a constant delay, is there any reason why we can't just use `k_sleep` or `k_busy_wait` instead of this?

Reply:
This is a calibrated delay loop for nanosecond-scale delays. It should be finely tunable and is controlled by a devicetree parameter. E.g. on esp32-c3, if the delay is zero, then I'm able to generate a 2.2 MHz square wave. If it's set to something non-zero (I can't remember exactly atm), then it can be tuned to generate a 1 MHz square wave. The esp32-c3 has an 80 MHz clock. On faster systems, this becomes more critical to ensure that bit-banging is actually slowed-down enough. 

k_sleep invokes the scheduler, so that's out of the question, and k_busy_wait works in units of microseconds using ticks, so it's also out of the question.

I did use k_busy_wait below to "wait at least several hundred nanoseconds" which gets rounded up to microsecond granularity, but that's far too coarse here.

Reply:
ack, could you please elaborate this a little bit more in the dts docs or add some comments? Currently the usage description is not very clear

Reply:
It was already described in the DTS bindings, but I've added some additional comments in the source.